pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -284,27 +290,45 @@ func (s *NSXServiceAccountService) getProxyEndpoints(ctx context.Context) (v1alp
 	return proxyEndpoints, nil
 }
 
-func (s *NSXServiceAccountService) DeleteNSXServiceAccount(ctx context.Context, namespacedName types.NamespacedName) error {
+func (s *NSXServiceAccountService) DeleteNSXServiceAccount(ctx context.Context, namespacedName types.NamespacedName, uid types.UID) error {
+	isDeleteSecret := false
+	nsxsa := &v1alpha1.NSXServiceAccount{}
+	if err := s.Client.Get(ctx, namespacedName, nsxsa); err != nil {
+		isDeleteSecret = true
+	} else if uid == nsxsa.UID {
+		isDeleteSecret = true
+	}
+
 	clusterName := s.getClusterName(namespacedName.Namespace, namespacedName.Name)
 	normalizedClusterName := util.NormalizeId(clusterName)
 	// delete Secret
-	secretName := namespacedName.Name + SecretSuffix
-	secretNamespace := namespacedName.Namespace
-	if err := s.Client.Delete(ctx, &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: secretNamespace}}); err != nil && !errors.IsNotFound(err) {
-		log.Error(err, "failed to delete", "secret", secretName, "namespace", secretNamespace)
-		return err
+	if isDeleteSecret {
+		secretName := namespacedName.Name + SecretSuffix
+		secretNamespace := namespacedName.Namespace
+		if err := s.Client.Delete(ctx, &v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: secretNamespace}}); err != nil && !errors.IsNotFound(err) {
+			log.Error(err, "failed to delete", "secret", secretName, "namespace", secretNamespace)
+			return err
+		}
 	}
 
+	isDeleteCCP := true
+	isDeletePI := true
+	if !isDeleteSecret {
+		isDeletePI = len(s.PrincipalIdentityStore.GetByIndex(common.TagScopeNSXServiceAccountCRUID, string(uid))) > 0
+		isDeleteCCP = len(s.ClusterControlPlaneStore.GetByIndex(common.TagScopeNSXServiceAccountCRUID, string(uid))) > 0
+	}
 	// delete ClusterControlPlane
-	cascade := true
-	if err := s.NSXClient.ClusterControlPlanesClient.Delete(siteId, enforcementpointId, normalizedClusterName, &cascade); err != nil {
-		log.Error(err, "failed to delete", "ClusterControlPlane", normalizedClusterName)
-		return err
+	if isDeleteCCP {
```

Comment:
looks like isDeleteCCP is always true, may I know the reason to have a if here?

Reply:
No, when `isDeleteSecret` is false, we need to check the UID of CCP/PI we want to delete to avoid unexpected deletion.